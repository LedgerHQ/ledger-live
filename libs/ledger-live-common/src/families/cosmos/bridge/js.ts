import createTransaction from "../js-createTransaction";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import getTransactionStatus from "../js-getTransactionStatus";
import prepareTransaction from "../js-prepareTransaction";
import signOperation from "../js-signOperation";
import { sync, scanAccounts } from "../js-synchronisation";
import updateTransaction from "../js-updateTransaction";
import type { CosmosValidatorItem, Transaction } from "../types";
import cosmosValidatorsManager from "../validators";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { defaultCosmosAPI } from "../api/Cosmos";
import {
  asSafeCosmosPreloadData,
  setCosmosPreloadData,
} from "../preloadedData";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const getPreloadStrategy = (_currency) => ({
  preloadMaxAge: 30 * 1000,
});

const currencyBridge: CurrencyBridge = {
  getPreloadStrategy,
  preload: async () => {
    const validators = await cosmosValidatorsManager.getValidators();
    setCosmosPreloadData({
      validators,
    });
    return Promise.resolve({
      validators,
    });
  },
  hydrate: (data: { validators?: CosmosValidatorItem[] }) => {
    if (!data || typeof data !== "object") return;
    const { validators } = data;
    if (
      !validators ||
      typeof validators !== "object" ||
      !Array.isArray(validators)
    )
      return;
    cosmosValidatorsManager.hydrateValidators(validators);
    setCosmosPreloadData(asSafeCosmosPreloadData(data));
  },
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  estimateMaxSpendable,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast: defaultCosmosAPI.broadcast,
};

export default {
  currencyBridge,
  accountBridge,
};
