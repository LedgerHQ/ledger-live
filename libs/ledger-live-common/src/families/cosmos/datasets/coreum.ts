import { CurrenciesData } from "@ledgerhq/types-live";
import type { CosmosAccountRaw, Transaction } from "../types";

const dataset: CurrenciesData<Transaction> = {
  FIXME_ignoreOperationFields: ["gas"],
  FIXME_ignoreAccountFields: ["cosmosResources", "operationsCount", "operations"],
  scanAccounts: [
    {
      name: "coreum seed 1",
      apdus: `
      => 550400001904636f72652c00008076000080000000800000000000000000
      <= 03633639c689420031c73c1157975ea734e22eae588045a96fc9495d28aba3f0d1636f726531306c366833717730357537716475716761666a38776c727833666a68723835327a3772356c329000
      => 550400001904636f72652c00008076000080000000800000000000000000
      <= 03633639c689420031c73c1157975ea734e22eae588045a96fc9495d28aba3f0d1636f726531306c366833717730357537716475716761666a38776c727833666a68723835327a3772356c329000
      => 550400001904636f72652c00008076000080010000800000000000000000
      <= 03da7e01e4a9c30aecb7574bda681dd021745f9982b3644b7733628f7579c8c310636f72653173656c3979733861376a6c736d727870393639327a666a6e636e6d68786b77733071723367399000
      => 550400001904636f72652c00008076000080020000800000000000000000
      <= 0209796438379822529674761b7a65e7b0262ba571a8bc8b1786ef942622b73bc7636f7265316d646a366a38396576347966656c76336b716e3770726a346839326572357a6c78737378776b9000
      `,
    },
  ],
  accounts: [
    {
      FIXME_tests: ["balance is sum of ops"],
      raw: {
        id: "js:2:coreum:core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9:",
        seedIdentifier: "03633639c689420031c73c1157975ea734e22eae588045a96fc9495d28aba3f0d1",
        name: "Coreum 2",
        starred: false,
        used: true,
        derivationMode: "",
        index: 1,
        freshAddress: "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
        freshAddressPath: "44'/118'/1'/0/0",
        freshAddresses: [
          {
            address: "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
            derivationPath: "44'/118'/1'/0/0",
          },
        ],
        blockHeight: 6313741,
        creationDate: "2023-07-26T02:02:57.000Z",
        operationsCount: 1,
        operations: [
          {
            id: "js:2:coreum:core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9:-0B17FF19A1E8DC703A3AEC457B414DE88845DD76BB4D95F8F6368F68D59248DF-IN",
            hash: "0B17FF19A1E8DC703A3AEC457B414DE88845DD76BB4D95F8F6368F68D59248DF",
            type: "IN",
            senders: ["core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2"],
            recipients: ["core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9"],
            accountId: "js:2:coreum:core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9:",
            blockHash: null,
            blockHeight: "6306437",
            extra: {
              validators: [],
              memo: "lb",
            },
            date: "2023-07-26T02:02:57.000Z",
            value: "1000000",
            fee: "8880",
            transactionSequenceNumber: 0,
          },
        ],
        pendingOperations: [],
        currencyId: "coreum",
        unitMagnitude: 6,
        lastSyncDate: "2023-07-26T05:46:40.684Z",
        balance: "1000000",
        spendableBalance: "1000000",
        balanceHistoryCache: {
          HOUR: {
            balances: [0, 1000000, 1000000, 1000000],
            latestDate: 1690347600000,
          },
          DAY: {
            balances: [0],
            latestDate: 1690322400000,
          },
          WEEK: {
            balances: [0],
            latestDate: 1690063200000,
          },
        },
        xpub: "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
        cosmosResources: {
          delegations: [],
          redelegations: [],
          unbondings: [],
          delegatedBalance: "0",
          pendingRewardsBalance: "0",
          unbondingBalance: "0",
          withdrawAddress: "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
        },
        swapHistory: [],
      } as unknown as CosmosAccountRaw,
    },
  ],
};

export default dataset;
