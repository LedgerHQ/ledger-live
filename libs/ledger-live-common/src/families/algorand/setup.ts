// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
