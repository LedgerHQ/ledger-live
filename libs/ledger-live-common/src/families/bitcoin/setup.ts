// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { firstValueFrom, from } from "rxjs";
import Transport from "@ledgerhq/hw-transport";
import { Bridge } from "@ledgerhq/types-live";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import Btc from "@ledgerhq/hw-app-btc";
import { createBridges } from "@ledgerhq/coin-bitcoin/bridge/js";
import type { SignerContext } from "@ledgerhq/coin-bitcoin/signer";
import makeCliTools from "@ledgerhq/coin-bitcoin/cli-transaction";
import bitcoinResolver from "@ledgerhq/coin-bitcoin/hw-getAddress";
import { signMessage } from "@ledgerhq/coin-bitcoin/hw-signMessage";
import { Transaction } from "@ledgerhq/coin-bitcoin/types";
import { GetAddressOptions, Resolver } from "../../hw/getAddress/types";
import { withDevice } from "../../hw/deviceAccess";
import { startSpan } from "../../performance";
import { GetAddressFn } from "@ledgerhq/coin-framework/bridge/getAddressWrapper";
import { getCurrencyConfiguration } from "../../config";
import { BitcoinConfigInfo } from "@ledgerhq/coin-bitcoin/lib/config";
import { SignMessage } from "../../hw/signMessage/types";

const createSigner = (transport: Transport, currency: CryptoCurrency) => {
  return new Btc({ transport, currency: currency.id });
};

const signerContext: SignerContext = <T>(
  deviceId: string,
  crypto: CryptoCurrency,
  fn: (signer: Btc) => Promise<T>,
): Promise<T> =>
  firstValueFrom(
    withDevice(deviceId)((transport: Transport) => from(fn(createSigner(transport, crypto)))),
  );

const getCurrencyConfig = (currency: CryptoCurrency) => {
  return { info: getCurrencyConfiguration<BitcoinConfigInfo>(currency) };
};

const perfLogger = {
  startSpan,
};

const bridge: Bridge<Transaction> = createBridges(signerContext, perfLogger, getCurrencyConfig);

export function createMessageSigner(): SignMessage {
  return (transport, account, messageData) => {
    const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
    return signMessage(signerContext)("", account, messageData);
  };
}

const messageSigner = {
  signMessage: createMessageSigner(),
};

const resolver: Resolver = (
  transport: Transport,
  addressOpt: GetAddressOptions,
): ReturnType<GetAddressFn> => {
  const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
  return bitcoinResolver(signerContext)("", addressOpt);
};

const cliTools = makeCliTools();

export { bridge, cliTools, resolver, messageSigner, signerContext };
