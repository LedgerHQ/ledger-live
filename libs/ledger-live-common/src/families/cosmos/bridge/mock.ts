import { BigNumber } from "bignumber.js";
import {
  NotEnoughBalance,
  RecipientRequired,
  InvalidAddress,
  FeeTooHigh,
} from "@ledgerhq/errors";
import type {
  CosmosAccount,
  CosmosValidatorItem,
  StatusErrorMap,
  Transaction,
} from "../types";
import {
  scanAccounts,
  signOperation,
  broadcast,
  sync,
  isInvalidRecipient,
} from "../../../bridge/mockHelpers";
import {
  setCosmosPreloadData,
  asSafeCosmosPreloadData,
} from "../preloadedData";
import { getMainAccount } from "../../../account";
import mockPreloadedData from "../preloadedData.mock";
import { makeAccountBridgeReceive } from "../../../bridge/mockHelpers";
import type {
  Account,
  AccountBridge,
  CurrencyBridge,
} from "@ledgerhq/types-live";
const receive = makeAccountBridgeReceive();

const defaultGetFees = (a, t) =>
  (t.fees || new BigNumber(0)).times(t.gas || new BigNumber(0));

const createTransaction = (): Transaction => ({
  family: "cosmos",
  mode: "send",
  amount: new BigNumber(0),
  recipient: "",
  fees: null,
  gas: null,
  memo: null,
  validators: [],
  sourceValidator: null,
  networkInfo: null,
  useAllAmount: false,
});

const updateTransaction = (t, patch) => ({ ...t, ...patch });

const estimateMaxSpendable = ({ account, parentAccount, transaction }) => {
  const mainAccount = getMainAccount(account, parentAccount);
  const estimatedFees = transaction
    ? defaultGetFees(mainAccount, transaction)
    : new BigNumber(5000);
  return Promise.resolve(
    BigNumber.max(0, account.balance.minus(estimatedFees))
  );
};

const getTransactionStatus = (account: Account, t: Transaction) => {
  const errors: StatusErrorMap = {};
  const warnings: StatusErrorMap = {};
  const useAllAmount = !!t.useAllAmount;
  const estimatedFees = defaultGetFees(account, t);
  const totalSpent = useAllAmount
    ? account.balance
    : new BigNumber(t.amount).plus(estimatedFees);
  const amount = useAllAmount
    ? account.balance.minus(estimatedFees)
    : new BigNumber(t.amount);

  if (amount.gt(0) && estimatedFees.times(10).gt(amount)) {
    warnings.feeTooHigh = new FeeTooHigh();
  }

  // Fill up transaction errors...
  if (totalSpent.gt(account.balance)) {
    errors.amount = new NotEnoughBalance();
  }

  // Fill up recipient errors...
  if (!t.recipient) {
    errors.recipient = new RecipientRequired("");
  } else if (isInvalidRecipient(t.recipient)) {
    errors.recipient = new InvalidAddress("");
  }

  return Promise.resolve({
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  });
};

const prepareTransaction = async (
  a: CosmosAccount,
  t: Transaction
): Promise<Transaction> => {
  if (!t.networkInfo) {
    return {
      ...t,
      gas: new BigNumber(1),
      fees: new BigNumber(500),
      networkInfo: {
        family: "cosmos",
        fees: new BigNumber(500),
      },
    };
  }

  return t;
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: () => {
    setCosmosPreloadData(mockPreloadedData);
    return Promise.resolve(mockPreloadedData);
  },
  hydrate: (data: { validators?: CosmosValidatorItem[] }) => {
    setCosmosPreloadData(asSafeCosmosPreloadData(data));
  },
};
export default {
  currencyBridge,
  accountBridge,
};
