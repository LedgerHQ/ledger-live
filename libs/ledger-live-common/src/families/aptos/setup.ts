// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
