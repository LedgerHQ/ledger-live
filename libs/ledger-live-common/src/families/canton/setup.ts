// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-canton/bridge/index";
import { getEnv } from "@ledgerhq/live-env";
import Transport from "@ledgerhq/hw-transport";
import { CantonSigner } from "@ledgerhq/coin-canton";
import cantonResolver from "@ledgerhq/coin-canton/signer";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-canton/test/cli";
import { CantonCoinConfig } from "@ledgerhq/coin-canton/config";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-canton/types";
import { LegacySignerCanton } from "@ledgerhq/live-signer-canton";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import cantonBridgeMock from "./bridge/mock";
import { Resolver } from "../../hw/getAddress/types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "../../currencies";

const createSigner: CreateSigner<CantonSigner> = (transport: Transport) => {
  return new LegacySignerCanton(transport);
};

const getCurrencyConfig = () => {
  // Use devnet for development/testing
  const currencyId = getEnv("MOCK") ? "canton_network_devnet" : "canton_network_devnet";
  return getCurrencyConfiguration<CantonCoinConfig>(getCryptoCurrencyById(currencyId));
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = getEnv("MOCK")
  ? cantonBridgeMock
  : createBridges(executeWithSigner(createSigner), getCurrencyConfig);

const resolver: Resolver = createResolver(createSigner, cantonResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
