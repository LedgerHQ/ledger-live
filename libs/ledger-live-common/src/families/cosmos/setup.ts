// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-cosmos/bridge/index";
import makeCliTools from "@ledgerhq/coin-cosmos/cli";
import { CosmosCoinConfig } from "@ledgerhq/coin-cosmos/config";
import cosmosResolver from "@ledgerhq/coin-cosmos/hw-getAddress";
import { CosmosAccount, Transaction, TransactionStatus } from "@ledgerhq/coin-cosmos/types/index";
import { CosmosSigner } from "@ledgerhq/coin-cosmos/types/signer";
import Cosmos from "@ledgerhq/hw-app-cosmos";
import Transport from "@ledgerhq/hw-transport";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import type { Bridge } from "@ledgerhq/types-live";
import { CosmosApp } from "@zondax/ledger-cosmos-js";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { getCurrencyConfiguration } from "../../config";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<CosmosSigner> = (transport: Transport) => {
  const cosmos = new CosmosApp(transport);
  const hwCosmos = new Cosmos(transport);
  return {
    getAddressAndPubKey: cosmos.getAddressAndPubKey.bind(cosmos),
    sign: cosmos.sign.bind(cosmos),
    getAddress: hwCosmos.getAddress.bind(hwCosmos),
  };
};

const getCurrencyConfig = (currency?: CryptoCurrency) => {
  if (!currency) {
    throw new Error("No currency provided");
  }
  return getCurrencyConfiguration<CosmosCoinConfig>(currency);
};
const bridge: Bridge<Transaction, CosmosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
  getCurrencyConfig,
);

const resolver: Resolver = createResolver(createSigner, cosmosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
