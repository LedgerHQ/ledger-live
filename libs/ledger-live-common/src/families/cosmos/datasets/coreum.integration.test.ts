import { DatasetTest } from "@ledgerhq/types-live";
import { testBridge } from "../../../__tests__/test-helpers/bridge";
import "../../../__tests__/test-helpers/setup";
import coreum from "./coreum";
import type { Transaction } from "../types";

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    coreum,
  },
};

testBridge(dataset);
