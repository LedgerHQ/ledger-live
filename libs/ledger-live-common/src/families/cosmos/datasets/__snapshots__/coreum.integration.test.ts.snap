// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coreum currency bridge scanAccounts coreum seed 1 1`] = `
Array [
  Object {
    "balance": "966840",
    "currencyId": "coreum",
    "derivationMode": "",
    "freshAddress": "core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2",
    "freshAddressPath": "44'/118'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2",
        "derivationPath": "44'/118'/0'/0/0",
      },
    ],
    "id": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:",
    "index": 0,
    "name": "Coreum 1",
    "nfts": undefined,
    "pendingOperations": Array [],
    "seedIdentifier": "03633639c689420031c73c1157975ea734e22eae588045a96fc9495d28aba3f0d1",
    "spendableBalance": "666840",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 6,
    "used": true,
    "xpub": "core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2",
  },
  Object {
    "balance": "1000000",
    "currencyId": "coreum",
    "derivationMode": "",
    "freshAddress": "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
    "freshAddressPath": "44'/118'/1'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
        "derivationPath": "44'/118'/1'/0/0",
      },
    ],
    "id": "js:2:coreum:core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9:",
    "index": 1,
    "name": "Coreum 2",
    "nfts": undefined,
    "pendingOperations": Array [],
    "seedIdentifier": "03633639c689420031c73c1157975ea734e22eae588045a96fc9495d28aba3f0d1",
    "spendableBalance": "1000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 6,
    "used": true,
    "xpub": "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
  },
  Object {
    "balance": "0",
    "currencyId": "coreum",
    "derivationMode": "",
    "freshAddress": "core1mdj6j89ev4yfelv3kqn7prj4h92er5zlxssxwk",
    "freshAddressPath": "44'/118'/2'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "core1mdj6j89ev4yfelv3kqn7prj4h92er5zlxssxwk",
        "derivationPath": "44'/118'/2'/0/0",
      },
    ],
    "id": "js:2:coreum:core1mdj6j89ev4yfelv3kqn7prj4h92er5zlxssxwk:",
    "index": 2,
    "name": "Coreum 3",
    "nfts": undefined,
    "pendingOperations": Array [],
    "seedIdentifier": "03633639c689420031c73c1157975ea734e22eae588045a96fc9495d28aba3f0d1",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 6,
    "used": false,
    "xpub": "core1mdj6j89ev4yfelv3kqn7prj4h92er5zlxssxwk",
  },
]
`;

exports[`coreum currency bridge scanAccounts coreum seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:",
      "blockHash": null,
      "blockHeight": "6306505",
      "contract": undefined,
      "extra": Object {
        "memo": "Ledger Live",
        "validators": Array [
          Object {
            "address": "corevaloper1ll9gdh5ur6gyv6swgshcm4zkkw4ttakt0zghmr",
            "amount": "300000",
          },
        ],
      },
      "fee": "14280",
      "hash": "08303E7722D24EA4660DF93EA7170090AAAD4B884284FFE6EE8CA67C9B0D1957",
      "id": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:-08303E7722D24EA4660DF93EA7170090AAAD4B884284FFE6EE8CA67C9B0D1957-DELEGATE",
      "operator": undefined,
      "recipients": Array [],
      "senders": Array [],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 1,
      "type": "DELEGATE",
      "value": "14280",
    },
    Object {
      "accountId": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:",
      "blockHash": null,
      "blockHeight": "6306437",
      "contract": undefined,
      "extra": Object {
        "memo": "lb",
        "validators": Array [],
      },
      "fee": "8880",
      "hash": "0B17FF19A1E8DC703A3AEC457B414DE88845DD76BB4D95F8F6368F68D59248DF",
      "id": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:-0B17FF19A1E8DC703A3AEC457B414DE88845DD76BB4D95F8F6368F68D59248DF-OUT",
      "operator": undefined,
      "recipients": Array [
        "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
      ],
      "senders": Array [
        "core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 0,
      "type": "OUT",
      "value": "1008880",
    },
    Object {
      "accountId": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:",
      "blockHash": null,
      "blockHeight": "6306340",
      "contract": undefined,
      "extra": Object {
        "memo": "Lc",
        "validators": Array [],
      },
      "fee": "10000",
      "hash": "B128DC8C3D820F0832BEBD81045FA65ABFDC88FB83246B37E3F1298AF3DE778A",
      "id": "js:2:coreum:core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2:-B128DC8C3D820F0832BEBD81045FA65ABFDC88FB83246B37E3F1298AF3DE778A-IN",
      "operator": undefined,
      "recipients": Array [
        "core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2",
      ],
      "senders": Array [
        "core1lptvx486j7630632q3ah0dwq20ghdrpxm3la9n",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 0,
      "type": "IN",
      "value": "1990000",
    },
  ],
  Array [
    Object {
      "accountId": "js:2:coreum:core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9:",
      "blockHash": null,
      "blockHeight": "6306437",
      "contract": undefined,
      "extra": Object {
        "memo": "lb",
        "validators": Array [],
      },
      "fee": "8880",
      "hash": "0B17FF19A1E8DC703A3AEC457B414DE88845DD76BB4D95F8F6368F68D59248DF",
      "id": "js:2:coreum:core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9:-0B17FF19A1E8DC703A3AEC457B414DE88845DD76BB4D95F8F6368F68D59248DF-IN",
      "operator": undefined,
      "recipients": Array [
        "core1sel9ys8a7jlsmrxp9692zfjncnmhxkws0qr3g9",
      ],
      "senders": Array [
        "core10l6h3qw05u7qduqgafj8wlrx3fjhr852z7r5l2",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 0,
      "type": "IN",
      "value": "1000000",
    },
  ],
  Array [],
]
`;
